File: JS/Trees.js
class Node {
  constructor (value) {
    this.value = value
    this.left = null
    this.right = null
  }
}

class BinarSearchTree {
  constructor () {
    this.root = null
  }

  find (val) {
    let current = this.root
    if (!current) {
      return false
    } else {
      while (true) {
        if (current.value === val) {
          return true
        } else if (val > current.value) {
          if (!current.right) {
            return false
          } else {
            current = current.right
          }
        } else if (val < current.value) {
          if (!current.left) {
            return false
          } else {
            current = current.left
          }
        } else {
          return false
        }
      }
    }
  }

  insert (val) {
    const newNode = new Node(val)

    if (!this.root) {
      this.root = newNode
      return this
    } else {
      let current = this.root
      while (true) {
        if (val === current.value) {
          return undefined
        }
        if (val > current.value) {
          if (current.right) {
            current = current.right
          } else {
            current.right = newNode
            return this
          }
        } else {
          if (val < current.value) {
            if (current.left) {
              current = current.left
            } else {
              current.left = newNode
              return this
            }
          }
        }
      }
    }
  }

  BFS () {
    const queue = []
    const visited = []
    let node = this.root

    if (!this.root) {
      return visited
    } else {
      queue.push(node)
      while (queue.length) {
        node = queue.shift()
        visited.push(node.value)
        if (node.left) { queue.push(node.left) }
        if (node.right) { queue.push(node.right) }
      }
      return visited
    }
  }

  DFSPreOrder () {
    const data = []
    function helper (node) {
      data.push(node.value)
      if (node.left) helper(node.left)
      if (node.right) helper(node.right)
    }
    helper(this.root)
    return data
  }

  DFSPostOrder () {
    const data = []
    function helper (node) {
      if (node.left) helper(node.left)
      if (node.right) helper(node.right)
      data.push(node.value)
    }
    helper(this.root)
    return data
  }

  DFSInOrder () {
    const data = []
    function helper (node) {
      if (node.left) helper(node.left)
      data.push(node.value)
      if (node.right) helper(node.right)
    }
    helper(this.root)
    return data
  }
}

const tree = new BinarSearchTree()
tree.insert(10)
tree.insert(6)
tree.insert(15)
tree.insert(20)
tree.insert(3)
tree.insert(8)
console.log(tree.DFSInOrder())

File: codinguniversity/python/trees.py
import queue

class Node:
    def __init__(self, val):
        self.val = val
        self.right = None
        self.left = None

    def __str__(self):
        return str(self.val)

    def __repr__(self):
        return str(self.val)

class BinarySearchTree:
    def __init__(self):
       self.head = None
    
    def __str__(self):
        #DFS
        # inorder gives a sorted list
        '''
        Time complexity O(n)
        Space complexity : O(h)
        '''
        arr = []
        def dfs(head):
            if not head:
                return            
            dfs(head.left)   
            arr.append(head.val)         
            dfs(head.right) 

        dfs(self.head)
        return str(arr)

    # def  get_height(self):

    def bfs(self):
        '''
        Time complexity O(n)
        Space complexity : best O(1) worst O(n/2)
        '''
        if not self.head:
            return None
        arr = []
        q = queue.Queue()
        q.put(self.head)
        while not q.empty():
            curr = q.get()
            arr.append(curr.val)
            if curr.left:
                q.put(curr.left)
            if curr.right:
                q.put(curr.right)
        return arr
            


    def insert(self, val):
        newNode = Node(val)
        if not self.head:
            self.head = newNode
            return self.head
        else:
            curr = self.head
            while curr:
                if val < curr.val:             
                    if curr.left:  
                        curr = curr.left
                    else:
                        curr.left = newNode
                        break
                else:
                    if curr.right:
                        curr = curr.right
                    else:
                        curr.right = newNode
                        break
        return self.head.left

    def find_height(self):
        def height(node):
            if not node:
                return -1
            return max(height(node.left), height(node.right)) + 1
        return height(self.head)

    def is_binary_search_tree(self):
        '''
        Time Complexity O(n)
        '''
        def helper(node, minVal, maxVal):
            # data has to always be within the range of mix and max
            if not node:
                return True
            if (node.val > minVal and node.val < maxVal
            and helper(node.left, minVal, node.val)
            and helper(node.right, node.val, maxVal)
            ):
                return True
            else:
                return False
        return helper(self.head, float('-inf'), float('inf'))
        


                

bst = BinarySearchTree()
bst.insert(5)
bst.insert(3)
bst.insert(4)
bst.insert(6)
print(bst.is_binary_search_tree())